contracts/src/FeeContract.sol

```diff
@@ -0,0 +1,98 @@
+
+pragma solidity ^0.8.0;
+// import { console } from "forge-std/console.sol";
+
+import { OwnableUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import { UUPSUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+contract FeeContract is Initializable, OwnableUpgradeable, UUPSUpgradeable {
+    // === Events ===
+    //
+    /// @notice Notify a new deposit
+    event Deposit(address indexed user, uint256 amount);
+    event Log(string func, uint256 gas);
+
+    /// @notice upgrade event when the proxy updates the implementation it's pointing to
+    event Upgrade(address implementation);
+    // === Constants ===
+    //
+    /// @notice max amount allowed to be deposited to prevent fat finger errors
+    // @TODO confirm this amount with product
+
+    uint256 public immutable MAX_DEPOSIT_AMOUNT = 1 ether;
+    uint256 public immutable MIN_DEPOSIT_AMOUNT = 0.1 ether;
+
+    // === Errors ===
+    //
+    /// @notice error types
+    error InvalidUserAddress();
+    error NonZeroDepositAmount();
+    error NotEnoughDeposited();
+    error DepositTooLarge();
+    error DepositTooSmall();
+    error FunctionDoesNotExist();
+    error NoFunctionCalled();
+
+    /// @notice store user balances in a mapping
+    mapping(address user => uint256 amount) public balances;
+
+    /// @notice since the constuctor initializes storage on this contract we disable it
+    /// @dev storage is on the proxy contract since it calls this contract via delegatecall
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @notice This contract is called by the proxy when you deploy this contract
+    function initialize() public initializer {
+        __Ownable_init(msg.sender); //sets owner to msg.sender
+        __UUPSUpgradeable_init();
+    }
+
+    /// @notice Revert if a method name does not exist
+    fallback() external payable {
+        // send / transfer (forwards 2300 gas to this fallback function)
+        // call (forwards all of the gas)
+        revert FunctionDoesNotExist();
+    }
+
+    /// @notice Revert if no method name was called
+    receive() external payable {
+        revert NoFunctionCalled();
+    }
+
+    /// @notice Allows anyone to deposit an ETH balance for any user
+    /// @dev the deposit amount is less than a specified threshold to prevent fat finger errors
+    function deposit(address user) public payable {
+        if (msg.value == 0) {
+            revert NonZeroDepositAmount();
+        }
+        if (msg.value < MIN_DEPOSIT_AMOUNT) {
+            revert DepositTooSmall();
+        }
+        if (msg.value > MAX_DEPOSIT_AMOUNT) {
+            revert DepositTooLarge();
+        }
+        if (user == address(0)) {
+            revert InvalidUserAddress();
+        }
+        balances[user] += msg.value;
```

Comment:
I think we do not need to track the balances of each user in the contract as the accounting will happen inside HotShot. Is the `balances` mapping for testing purposes only?

Reply:
Oh I see, but would hotshot need to read the balance deposited from this contract initially? Also when the user deposits more

Reply:
any thoughts on this @alxiong @jbearer ?

Reply:
While it is possible, I don't think it is needed as it should be enough to read from the deposit events. That said it depends on the implementation of the fee accounting inside HotShot cc @jbearer.

Reply:
Anyways, this function cannot create any harm and makes testing simpler so I am ok to keep it.

Reply:
very true @ being able to read from the deposit events. Right now, we assume that Hotshot takes care of all the accounting and also storage. So maybe it's worth considering removing it so that the L1 smart contract storage doesn't increase which can impact gas fees over time. Unless we want to consider things like onchain (Ethereum L1) integrity and a possible future where we handle ERC20 tokens and then give the users a wrapped version that's represented in their wallet?

Reply:
personally, i prefer to keep the map in contract, it's not _very_ expensive as it's minimally simple.

having this map makes upgrade with "less" trust, instead of completely at the arbitration of the admin (i'd imagine ppl trusting Espresso, but not when there's many millions deposited and our admin and just "oops, I upgrade a contract that can claim a different mapping, users can prove historical event to claim their coin if they are so keen")